File: src/main/java/com/sf_lolitahag/hanabi/AbsPaintArray.java
/*
Copyright (C) 2016 Koji Wakamiya
Copyright (C) 2016 Masanori Kojima
Released under the MIT license
http://opensource.org/licenses/mit-license.php
*/
package com.sf_lolitahag.hanabi;

import java.awt.Graphics;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.stream.IntStream;

public abstract class AbsPaintArray {

  private final List<PaintObject> paintList = new CopyOnWriteArrayList<>();

  public void init(int x, int y) {
    paintList.clear();
  }

  public int getTopX() {
    return paintList.get(0).getX();
  }

  public int getTopY() {
    return paintList.get(0).getY();
  }

  public int getTopAlpha() {
    return paintList.get(0).getColor().getAlpha();
  }

  public void draw(final Graphics g) {
    synchronized (paintList) {
      paintList.forEach(paintObject -> drawObject(g, paintObject));
    }
  }

  protected void initList(int x, int y, int r, int g, int b, int alpha, int tail) {
    initList(x, y, r, g, b, alpha, tail, 0, 0);
  }

  protected void initList(int x, int y, int r, int g, int b, int alpha, int tail, int moveX,
      int moveY) {
    IntStream.range(0, tail).forEach(index ->
        paintList.add(new PaintObjectImpl(x, y, r, g, b, alpha, tail, index, moveX, moveY)));
  }

  protected void addTop() {
    synchronized (paintList) {
      PaintObject newObject = paintList.get(0).clone();
      newObject.movePositionNext();

      paintList.parallelStream().forEach(PaintObject::moveIndexNext);
      paintList.add(0, newObject);
      paintList.remove(paintList.size() - 1);
    }
  }

  protected void addTop(int newMoveX, int newMoveY) {
    synchronized (paintList) {
      PaintObject newObject = paintList.get(0).clone();
      newObject.setMoveX(newMoveX);
      newObject.setMoveY(newMoveY);
      newObject.movePositionNext();

      paintList.parallelStream().forEach(PaintObject::moveIndexNext);
      paintList.add(0, newObject);
      paintList.remove(paintList.size() - 1);
    }
  }

  protected void updateAlpha(int alpha) {
    synchronized (paintList) {
      paintList.parallelStream().forEach(object -> object.setAlpha(alpha));
    }
  }

  public abstract void next();

  protected abstract void drawObject(Graphics g, PaintObject object);
}

File: src/main/kotlin/com/sf_lolitahag/hanabi/AbsPaintArray.kt
/*
Copyright (C) 2016 Koji Wakamiya
Copyright (C) 2016 Masanori Kojima
Released under the MIT license
http://opensource.org/licenses/mit-license.php
*/
package com.sf_lolitahag.hanabi

import java.awt.Graphics
import java.util.concurrent.CopyOnWriteArrayList
import java.util.stream.IntStream

abstract class AbsPaintArray {

    private val paintList = CopyOnWriteArrayList<PaintObject>()

    val topX: Int
        get() = paintList[0].x

    val topY: Int
        get() = paintList[0].y

    val topAlpha: Int
        get() = paintList[0].color.alpha

    open fun init(x: Int, y: Int) {
        paintList.clear()
    }

    fun draw(g: Graphics) {
        synchronized(paintList) {
            paintList.forEach { paintObject -> drawObject(g, paintObject) }
        }
    }

    @JvmOverloads
    protected fun initList(x: Int, y: Int, r: Int, g: Int, b: Int, alpha: Int, tail: Int, moveX: Int = 0,
                           moveY: Int = 0) {
        IntStream.range(0, tail).forEach { index -> paintList.add(PaintObjectImpl(x, y, r, g, b, alpha, tail, index, moveX, moveY)) }
    }

    protected fun addTop() {
        synchronized(paintList) {
            val newObject = paintList[0].clone()
            newObject.movePositionNext()

            paintList.forEach { it.moveIndexNext() }
            paintList.add(0, newObject)
            paintList.removeAt(paintList.size - 1)
        }
    }

    protected fun addTop(newMoveX: Int, newMoveY: Int) {
        synchronized(paintList) {
            val newObject = paintList[0].clone()
            newObject.setMoveX(newMoveX)
            newObject.setMoveY(newMoveY)
            newObject.movePositionNext()

            paintList.forEach { it.moveIndexNext() }
            paintList.add(0, newObject)
            paintList.removeAt(paintList.size - 1)
        }
    }

    protected fun updateAlpha(alpha: Int) {
        synchronized(paintList) {
            paintList.parallelStream().forEach { `object` -> `object`.setAlpha(alpha) }
        }
    }

    abstract operator fun next()

    protected abstract fun drawObject(g: Graphics, `object`: PaintObject)
}
